### PR TITLE
Decomissioning on change plan

### PR DESCRIPTION
**Pull request by clairecfu**
- [x] decommissioning endpoint takes change_plan query param
- [x] asset get detail endpoint takes change_plan query param
- [x] decommissioned assets ignored from change plan get many assets
- [x] asset detail view shows that an asset has been decommissioned on change plan
- [x] updating related_decomissionned_asset field on AssetCP

TODO:
- decomissioned assets get many -> need to includes assets decommissioned on change plan @juliasaveliff

### CODE REVIEW

File: rackcity/models/asset.py [hunk 1]
```diff
@@ -17,12 +17,12 @@ def get_next_available_asset_number():
             return asset_number
 
 
-def get_assets_for_cp(change_plan=None):
+def get_assets_for_cp(change_plan, show_decommissioned=False):
```
Comment (clairecfu) [inline rackcity/models/asset.py]: made change_plan a required param, and also added optional param to return decomissioned  assets on cp

**Comment by clairecfu** [comment]
i pulled from my branch `claire-cp-detailview` because I needed some front end changes I had made on that branch. Should not affect any of the backend logic.

File: rackcity/models/asset.py [hunk 2]
```diff
@@ -31,7 +31,8 @@ def get_assets_for_cp(change_plan=None):
     for assetCP in assetsCP:
         if (assetCP.related_asset) and (assetCP.related_asset) in assets:
             assets = assets.filter(~Q(id=assetCP.related_asset.id))
-    # TODO remove any Asset or AssetCP that has been decommissioned in this CP
+    if not show_decommissioned:
+        assetsCP = AssetCP.objects.filter(change_plan=change_plan, is_decommissioned=False)
```
Comment (juliasaveliff) [inline rackcity/models/asset.py]: you can just apply the filter on the query set you already have
```
assetsCP = assetsCP.filter(is_decommissioned=False)
```
so you don't have to refilter on change_plan

File: rackcity/models/asset.py [hunk 3]
```diff
@@ -60,10 +62,11 @@ def validate_hostname_uniqueness(value, asset_id, change_plan, related_asset):
 
 def validate_asset_number_uniqueness(value, asset_id, change_plan, related_asset):
     assets, assets_cp = get_assets_for_cp(change_plan.id)
-    matching_assets = assets_cp.filter(asset_number=value)
+    matching_assets = assets_cp.filter(asset_number=value,change_plan=change_plan)
```
Comment (juliasaveliff) [inline rackcity/models/asset.py]: again assets_cp queryset is already filtered on change_plan so you shouldn't need that here - were you getting unexpected results?

File: rackcity/views/asset_views.py [hunk 4]
```diff
@@ -130,7 +130,7 @@ def asset_detail(request, id):
             return response
     try:
         if change_plan:
-            assets, assets_cp = get_assets_for_cp(change_plan.id)
+            assets, assets_cp = get_assets_for_cp(change_plan.id, show_decommissioned=True)
```
Comment (juliasaveliff) [inline rackcity/views/asset_views.py]: is the decommission view set on the front end using the is_decommissioned flag?

File: rackcity/views/decommission_views.py [hunk 5]
```diff
@@ -41,7 +48,60 @@ def decommission_asset(request):
             status=HTTPStatus.BAD_REQUEST
         )
     id = data['id']
+    if change_plan:
+        if not AssetCP.objects.filter(id=id).exists() and Asset.objects.filter(id=id).exists:
```
Comment (juliasaveliff) [inline rackcity/views/decommission_views.py]: should be `.exists()`?

File: rackcity/views/decommission_views.py [hunk 6]
```diff
@@ -41,7 +48,60 @@ def decommission_asset(request):
...
+                decommissioned_asset_cp = AssetCP.objects.get(id=id)
+                decommissioned_asset_cp.is_decommissioned=True
+
+            except Asset.DoesNotExist:
+                return JsonResponse(
+                    {
+                        "failure_message":
+                            Status.ERROR.value +
+                            "Asset" + GenericFailure.DOES_NOT_EXIST.value,
+                        "errors": "No existing asse in change plan with id="+str(id)
```
Comment (juliasaveliff) [inline rackcity/views/decommission_views.py]: typo

File: rackcity/views/decommission_views.py [hunk 7]
```diff
@@ -31,6 +32,12 @@ def decommission_asset(request):
     Decommission a live asset
     """
     data = JSONParser().parse(request)
+    change_plan = None
+
```
Comment (juliasaveliff) [inline rackcity/views/decommission_views.py]: you need to keep the asset permission check at the top of this view - and you will need to check permission on asset_cp if there's a change plan before you do anything else

Comment (juliasaveliff) [reply rackcity/views/decommission_views.py]: it's below your code changes in the view, starts with...
```
if not user_has_asset_permission(
        request.user,
        asset.rack.datacenter
    ):
```

File: rackcity/views/decommission_views.py [hunk 8]
```diff
@@ -99,7 +160,13 @@ def decommission_asset(request):
             status=HTTPStatus.BAD_REQUEST
         )
     else:
+        
+        for assetcp in AssetCP.objects.filter(related_asset=id):
+            print(assetcp)
+            assetcp.related_decommissioned_asset = decommissioned_asset_object
```
Comment (juliasaveliff) [inline rackcity/views/decommission_views.py]: do you need to mark as conflict or does that happen with signals on postsave?

File: rackcity/views/asset_views.py (hunk 4 above)
Comment (clairecfu) [reply rackcity/views/asset_views.py]: yeah -> if an asset has been decommissioned you still want to see the detailview page. I added a warning that shows that it has been decomissioned and disabled edit/delete buttons

File: rackcity/views/decommission_views.py (hunk 5 above)
Comment (clairecfu) [reply rackcity/views/decommission_views.py]: oops good catch

File: rackcity/views/decommission_views.py (hunk 7 above)
Comment (clairecfu) [reply rackcity/views/decommission_views.py]: moved the code up! do I need to additional work to check permission on asset_cp?

Comment (clairecfu) [reply rackcity/views/decommission_views.py]: i thought it was datacenter specific?

Comment (clairecfu) [reply rackcity/views/decommission_views.py]: jk i see what you mean, refactoring this method

File: rackcity/views/decommission_views.py (hunk 8 above)
Comment (clairecfu) [reply rackcity/views/decommission_views.py]: i thought we were doing a separate check? the is_conflict field is only if asset is modified or deleted. you would detect this conflict by just checking if the asset has a related_decomissioned_asset -> sorry i think i forgot to flag this on your PR

Comment (juliasaveliff) [reply rackcity/views/decommission_views.py]: oh right I forgot 👍

File: rackcity/views/decommission_views.py [hunk 9]
```diff
@@ -41,33 +48,104 @@ def decommission_asset(request):
...
+                        "failure_message":
+                            Status.ERROR.value +
+                            "Asset" + GenericFailure.DOES_NOT_EXIST.value,
+                        "errors": "No existing asset in change plan with id="+str(id)
+                    },
+                    status=HTTPStatus.BAD_REQUEST
+                )
+        if not user_has_asset_permission(
+            request.user,
+            decomissioned_asset_cp.rack.datacenter
```
Comment (juliasaveliff) [inline rackcity/views/decommission_views.py]: typo - `decomissioned`

File: rackcity/models/asset.py (hunk 3 above)
Comment (juliasaveliff) [reply rackcity/models/asset.py]: still wondering about this

File: rackcity/views/decommission_views.py (hunk 9 above)
Comment (clairecfu) [reply rackcity/views/decommission_views.py]: omg thank god i did not work on decommissioning

**Pull request merged**